hypnotoad/gui/gui.py

```diff
@@ -80,8 +84,8 @@ def set_clicked(widget, function):
 
         self.action_Quit.triggered.connect(self.close)
 
-        self.default_options = dict(tokamak.TokamakEquilibrium.default_options.items())
-        self.options = copy.deepcopy(self.default_options)
+        self.options = tokamak.TokamakEquilibrium.default_options.push({})
```

Comment:
Is this the `options` idiom for making a copy? I couldn't get `copy.deepcopy` to work with it

Reply:
`Options` probably doesn't like you making a copy (see after following digression).

`Options` may not have been the best choice, possibly over-flexible for what we need, but things I like about it:
- call the options like `self.options.orthogonal` instead of `self.options["orthogonal"]`
- available options and default values are implemented in just one place, the `default_options` class member of whatever class uses them - good for maintainability. Downside is that the default has to be a static value, which has led to some slightly clunky workarounds when we want to use the value of another option as the default.
- it mostly does what we need, and avoids needing to write/test/debug a custom options class.

The way it works (https://options.readthedocs.io/en/latest/) is that a 'new' instance of the options is not independent from the instance it's created from - you create a hierarchy (using a library called `chainmap` if I remember right) rather than a set of independent copies. The new instance can override any option, but any that are not explicitly set on the instance use the value from the parent - with the advantage that if the parent is changed _after_ the new instance is created, the new instance also uses the changed value. We don't really get any advantage from this because we don't have many layers of options, and in this PR I've made the gui's `self.options` just a reference to `self.eq.user_options` anyway, but it's probably why copies don't work nicely. The `push()` method is the one that says 'create a new layer' and is intended to be called with a `kwargs` `dict`. As illustration:
```
In [1]: from options import Options                                             

In [2]: parent = Options(a=1, b=2)                                              

In [3]: child = parent.push({"b": 4})                                           

In [4]: child.c = 3                                                             

In [5]: grandchild = child.push({})                                             

In [6]: parent                                                                  
Out[6]: Options(a=1, b=2)

In [7]: child                                                                   
Out[7]: Options(c=3, a=1, b=4)

In [8]: grandchild                                                              
Out[8]: Options(c=3, a=1, b=4)

In [9]: parent.a = 42                                                           

In [10]: parent                                                                 
Out[10]: Options(a=42, b=2)

In [11]: child                                                                  
Out[11]: Options(c=3, a=42, b=4)

In [12]: grandchild                                                             
Out[12]: Options(c=3, a=42, b=4)

In [13]: parent.b = 77                                                          

In [14]: parent                                                                 
Out[14]: Options(a=42, b=77)

In [15]: child                                                                  
Out[15]: Options(c=3, a=42, b=4)

In [16]: grandchild                                                             
Out[16]: Options(c=3, a=42, b=4)
```

Reply:
I just wanted to mention that `options` pulls in additional dependencies.

For your first point, that is only a few lines, I did that to drop the dependency on `Bunch`.

As `options` doesn't even seem to have a public bug tracker (same is true for some dependencies) I do worry about pulling that in ...

Reply:
There's a couple things about `options` that I don't like that much: not being able to `deepcopy`, removing keys from the passed-in object.

To make the options form in the gui more useful, I think we would want to add some additional features to the options class at any rate: keeping track of documentation, expected type of values, allowed values. For example, make `refine_method` into a drop-down list of the available methods.

It's not causing enough pain at the minute to warrant changing it, but I suspect in the long-term `options` might be more trouble than it's worth.

Reply:
All good points. If someone knows of or wants to write a replacement, it would be welcome! I'll make in issue to point to this discussion...